node/src/actors/chain_manager/handlers.rs

```diff
@@ -1409,6 +1408,81 @@ impl Handler<GetBalance> for ChainManager {
     }
 }
 
+impl Handler<GetSupplyInfo> for ChainManager {
+    type Result = Result<SupplyInfo, failure::Error>;
+
+    fn handle(&mut self, _msg: GetSupplyInfo, _ctx: &mut Self::Context) -> Self::Result {
+        if self.sm_state != StateMachine::Synced {
+            return Err(ChainManagerError::NotSynced {
+                current_state: self.sm_state,
+            }
+            .into());
+        }
+
+        let total_supply = 2_500_000_000_000_000_000;
+
+        let chain_info = self.chain_state.chain_info.as_ref().unwrap();
+        let halving_period = chain_info.consensus_constants.halving_period;
+        let initial_block_reward = chain_info.consensus_constants.initial_block_reward;
+        let collateral_minimum = chain_info.consensus_constants.collateral_minimum;
+
+        let epoch = self.current_epoch.unwrap();
+        let current_time = u64::try_from(get_timestamp()).unwrap();
+
+        let mut current_unlocked_supply = 0;
+        let mut current_locked_supply = 0;
+        for (_output_pointer, value_transfer_output) in self.chain_state.unspent_outputs_pool.iter()
+        {
+            if value_transfer_output.0.time_lock <= current_time {
+                current_unlocked_supply += value_transfer_output.0.value;
+            } else {
+                current_locked_supply += value_transfer_output.0.value;
+            }
+        }
+
+        let in_flight_requests = self
+            .chain_state
+            .data_request_pool
+            .data_request_pool
+            .len()
+            .try_into()
+            .unwrap();
+        let locked_wits_by_requests = self
+            .chain_state
+            .data_request_pool
+            .locked_wits_by_requests(collateral_minimum);
+
+        let (mut blocks_minted, mut blocks_minted_reward) = (0, 0);
+        let (mut blocks_missing, mut blocks_missing_reward) = (0, 0);
+        for e in 1..epoch {
+            let block_reward = block_reward(epoch, initial_block_reward, halving_period);
```

Comment:
I think there is a bug here, although probably we wouldn't have noticed before first halving: 
```suggestion
            let block_reward = block_reward(e, initial_block_reward, halving_period);
```

Let me suggest though to rename `epoch` to `current_epoch` and `e` to `epoch` so the iterator clearly reads `for epoch in 1..current_epoch`, which more clearly states what's going on and avoids this kind of confusion.